pkg/volume/cinder_local/cinder.go

```diff
@@ -0,0 +1,432 @@
+		cinderVolume: &cinderVolume{
+			manager: manager,
+			plugin:  plugin,
+		},
+		options: options,
+	}, nil
+}
+
+func (plugin *cinderLocalPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volume.Spec, error) {
+	mounter := plugin.host.GetMounter()
+	pluginDir := plugin.host.GetPluginDir(plugin.GetPluginName())
+	sourceName, err := mounter.GetDeviceNameFromMount(mountPath, pluginDir)
+	if err != nil {
+		return nil, fmt.Errorf("could not get device name from mount point: %v", err)
+	}
+	glog.V(4).Infof("Found volume %s mounted to %s", sourceName, mountPath)
+	cinderVolume := &api.Volume{
+		Name: volumeName,
+		VolumeSource: api.VolumeSource{
+			CinderLocal: &api.CinderLocalVolumeSource{
+				CinderVolumeSource: api.CinderVolumeSource{
+					VolumeID: sourceName,
+					FSType:   "",
+					ReadOnly: false,
+				},
+			},
+		},
+	}
+	return volume.NewSpecFromVolume(cinderVolume), nil
+}
+
+// Abstract interface to PD operations.
+type cdManager interface {
+	// Attaches the disk to the kubelet's host machine.
+	AttachDisk(mounter *cinderVolumeMounter, mntPoint string) error
+	// Detaches the disk from the kubelet's host machine.
+	DetachDisk(unmounter *cinderVolumeUnmounter, mntPoint string) error
+	// Creates a volume
+	CreateVolume(provisioner *cinderVolumeProvisioner) (volumeID string, volumeSizeGB int, secretRef string, err error)
+	// Deletes a volume
+	DeleteVolume(deleter *cinderVolumeDeleter) error
+}
+
+// cinderPersistentDisk volumes are disk resources provided by C3
+// that are attached to the kubelet's host machine and exposed to the pod.
+type cinderVolume struct {
+	volName string
+	// Specifies the partition to mount
+	//partition string
+	// Utility interface that provides API calls to the provider to attach/detach disks.
+	manager cdManager
+	// Mounter interface that provides system calls to mount the global path to the pod local path.
+	mounter mount.Interface
+	plugin  *cinderLocalPlugin
+	volume.MetricsNil
+}
+
+type cinderVolumeMounter struct {
+	*cinderVolume
+	readOnly bool
+	pod      *api.Pod
+	// Specifies the cinder volume details
+	source *api.CinderLocalVolumeSource
+}
+
+var _ volume.Mounter = &cinderVolumeMounter{}
+
+func detachDiskLogError(cd *cinderVolume, podUID types.UID, mntPoint string) {
+	u := cinderVolumeUnmounter{
+		cinderVolume: cd,
+		podUID:       podUID,
+	}
+	if err := cd.manager.DetachDisk(&u, mntPoint); err != nil {
+		glog.Warningf("Failed to detach disk: %v (%v)", cd, err)
+	}
+}
+
+func (m *cinderVolumeMounter) GetPath() string {
+	return m.getPath(m.pod.UID)
+}
+
+func (m *cinderVolumeMounter) GetAttributes() volume.Attributes {
+	return volume.Attributes{
+		ReadOnly:        m.readOnly,
+		Managed:         !m.readOnly,
+		SupportsSELinux: true,
+	}
+}
+
+func (m *cinderVolumeMounter) SetUp(fsGroup *int64) error {
+	return m.SetUpAt(m.GetPath(), fsGroup)
+}
+
+// SetUp bind mounts to the volume path.
+func (m *cinderVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
```

Comment:
Implement attach/detach interface and make it a attachable volume, see https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/cinder/attacher.go